zebra/zebra_vty.c

```diff
@@ -77,7 +78,183 @@ static void vty_show_ip_route_summary_prefix(struct vty *vty,
+
+	XFREE(MTYPE_STATIC_ROUTE, shr);
+}
+
+static int static_list_compare(void *arg1, void *arg2)
+{
+	struct static_hold_route *shr1 = arg1;
+	struct static_hold_route *shr2 = arg2;
+	int ret;
+
+	ret = strcmp(shr1->vrf_name, shr2->vrf_name);
+	if (ret)
+		return ret;
+
+	ret = strcmp(shr2->nhvrf_name, shr2->nhvrf_name);
+	if (ret)
+		return ret;
+
+	ret = shr1->afi - shr2->afi;
+	if (ret)
+		return ret;
+
+	ret = shr1->safi - shr2->afi;
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->dest_str, shr2->dest_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->mask_str, shr2->mask_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->src_str, shr2->src_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->gate_str, shr2->gate_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->ifname, shr2->ifname);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->flag_str, shr2->flag_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->tag_str, shr2->tag_str);
+	if (ret)
+		return ret;
+
+	ret = static_list_compare_helper(shr1->distance_str,
+					 shr2->distance_str);
+	if (ret)
+		return ret;
+
+	return static_list_compare_helper(shr1->label_str, shr2->label_str);
+}
+
+
 /* General function for static route. */
+static int zebra_static_route_holdem(
+	struct zebra_vrf *zvrf, struct zebra_vrf *nh_zvrf, afi_t afi,
+	safi_t safi, const char *negate, const char *dest_str,
+	const char *mask_str, const char *src_str, const char *gate_str,
+	const char *ifname, const char *flag_str, const char *tag_str,
+	const char *distance_str, const char *label_str)
+{
+	struct static_hold_route *shr, *lookup;
+	struct listnode *node;
+
+	shr = XCALLOC(MTYPE_STATIC_ROUTE, sizeof(*shr));
+	shr->vrf_name = XSTRDUP(MTYPE_STATIC_ROUTE, zvrf->vrf->name);
+	shr->nhvrf_name = XSTRDUP(MTYPE_STATIC_ROUTE, nh_zvrf->vrf->name);
+	shr->afi = afi;
+	shr->safi = safi;
+	if (dest_str)
+		shr->dest_str = XSTRDUP(MTYPE_STATIC_ROUTE, dest_str);
+	if (mask_str)
+		shr->mask_str = XSTRDUP(MTYPE_STATIC_ROUTE, mask_str);
+	if (src_str)
+		shr->src_str = XSTRDUP(MTYPE_STATIC_ROUTE, src_str);
+	if (gate_str)
+		shr->gate_str = XSTRDUP(MTYPE_STATIC_ROUTE, gate_str);
+	if (ifname)
+		shr->ifname = XSTRDUP(MTYPE_STATIC_ROUTE, ifname);
+	if (flag_str)
+		shr->flag_str = XSTRDUP(MTYPE_STATIC_ROUTE, flag_str);
+	if (tag_str)
+		shr->tag_str = XSTRDUP(MTYPE_STATIC_ROUTE, tag_str);
+	if (distance_str)
+		shr->distance_str = XSTRDUP(MTYPE_STATIC_ROUTE, distance_str);
+	if (label_str)
+		shr->label_str = XSTRDUP(MTYPE_STATIC_ROUTE, label_str);
+
+	for (ALL_LIST_ELEMENTS_RO(static_list, node, lookup)) {
+		if (static_list_compare(shr, lookup) == 0)
+			break;
+	}
+
+	if (lookup) {
+		if (negate) {
+			listnode_delete(static_list, lookup);
+			static_list_delete(shr);
+			static_list_delete(lookup);
+
+			return CMD_SUCCESS;
+		}
+
+		assert(0);
```

Comment:
I've seen a neat assert variant used elsewhere in the codebase that provides some debugging info, it looks like this:
`assert(!"my error message")`
It's helped me in the past with debugging. Alternatively a comment around this area might be good.

Reply:
fixed